File: examples/poi/02_create_and_verify_multi_target_proof.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! # Create and Verify a Multi-Target Proof
//!
//! A single Proof of Inclusion can authenticate several claims about the same
//! transaction. This example proves the transaction itself, one object changed
//! by the transaction, and one event emitted by the transaction.
//!
//! Combining related targets avoids duplicating the transaction and checkpoint
//! evidence across separate proofs. Every object and event target must belong to
//! the selected transaction; the builder rejects targets from another transaction.

use anyhow::{Context, Result, ensure};
use poi_examples::prepare_poi_example;
use poi_rs::{CommitteeResolution, Proof};

/// Demonstrates how to:
/// 1. Create transaction, object, and event targets in one execution.
/// 2. Construct one proof containing all three claims.
/// 3. Inspect the targets resolved by the builder.
/// 4. Transfer the proof through its portable JSON representation.
/// 5. Authenticate the checkpoint committee from the active network's genesis.
/// 6. Verify every target in one operation.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Proof of Inclusion: Create and Verify a Multi-Target Proof ===\n");

    let context = prepare_poi_example().await?;
    let targets = context.create_notarization("PoI multi-target example").await?;
    let transaction_digest = targets.transaction_digest;
    let object_id = targets.object_id;
    let event_id = targets.event_id;
    let client = &context.poi_client;

    println!("Network:            {}", context.network_alias);
    println!("Transaction target: {transaction_digest}");
    println!("Object target:      {object_id}");
    println!("Event target:       {transaction_digest}:{}\n", event_id.event_seq);

    // The explicit transaction and event identify one execution. The object ID
    // is resolved at the exact version recorded in that transaction's effects.
    let proof = client
        .proof()
        .transaction(transaction_digest)
        .object(object_id)
        .event(event_id)
        .build()
        .await
        .context("failed to construct the multi-target proof")?;

    ensure!(
        proof.targets().transaction == Some(transaction_digest),
        "the proof must contain the requested transaction target"
    );
    ensure!(
        proof.targets().objects.len() == 1,
        "the proof must contain one object target"
    );
    ensure!(
        proof.targets().objects[0].as_inner().object_ref().object_id == object_id,
        "the proof must contain the requested object target"
    );
    ensure!(
        proof.targets().events == [event_id],
        "the proof must contain the requested event target"
    );
    ensure!(
        proof.transaction_proof.events.is_some(),
        "event evidence must be present when an event is targeted"
    );

    println!("Proof constructed:");
    println!("  checkpoint epoch:  {}", proof.checkpoint_summary.epoch());
    println!("  checkpoint number: {}", proof.checkpoint_summary.sequence_number);
    println!("  object targets:    {}", proof.targets().objects.len());
    println!("  event targets:     {}\n", proof.targets().events.len());

    // A verifier receives the complete proof as untrusted input. JSON is used
    // here to model transfer across a file, API, message, or process boundary.
    let proof_json = proof.to_json_vec().context("failed to serialize the proof as JSON")?;
    println!("Serialized proof size: {} bytes", proof_json.len());
    let received_proof = Proof::from_json_slice(&proof_json).context("failed to deserialize the received proof")?;

    let genesis = context.load_genesis().await?;
    let resolution = CommitteeResolution::from_genesis(genesis)
        .context("failed to load the committee from the trusted genesis blob")?;
    let verifier = client.verifier(resolution);

    // Genesis-anchored resolution authenticates every preceding committee
    // transition. This may take a long time on an established network.
    verifier
        .verify(&received_proof)
        .await
        .context("multi-target proof verification failed")?;

    println!("\nMulti-target proof verified successfully.");
    println!("The transaction, changed object, and emitted event are authenticated by one proof.");

    Ok(())
}

File: bindings/wasm/poi_wasm/examples/src/02_create_and_verify_multi_target_proof.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * # Create and Verify a Multi-Target Proof
 *
 * A single Proof of Inclusion can authenticate several claims about the same
 * transaction. This example proves the transaction itself, one object changed
 * by it, and one event emitted by it.
 *
 * Combining related targets avoids duplicating transaction and checkpoint
 * evidence. Every object and event target must belong to the same transaction.
 */

import { strict as assert } from "node:assert";

import { fromBase58, fromHex, normalizeIotaObjectId } from "@iota/iota-sdk/utils";
import { CommitteeResolution, PoiClient, Proof } from "@iota/poi-wasm";
import {
    loadMainnetGenesis,
    MAINNET_TRANSACTION_DIGEST,
    STAKED_IOTA_OBJECT_ID,
    STAKING_REQUEST_EVENT_SEQUENCE,
} from "./util.js";

/** Demonstrates how to construct and verify one proof containing three related targets. */
export async function createAndVerifyMultiTargetProof(): Promise<void> {
    console.log("=== Proof of Inclusion: Create and Verify a Multi-Target Proof ===\n");

    const client = PoiClient.mainnet();
    const transaction = fromBase58(MAINNET_TRANSACTION_DIGEST);
    const object = fromHex(normalizeIotaObjectId(STAKED_IOTA_OBJECT_ID, false, true));

    console.log("Network:            mainnet");
    console.log(`Transaction target: ${MAINNET_TRANSACTION_DIGEST}`);
    console.log(`Object target:      ${STAKED_IOTA_OBJECT_ID}`);
    console.log(`Event target:       ${MAINNET_TRANSACTION_DIGEST}:${STAKING_REQUEST_EVENT_SEQUENCE}\n`);

    // The explicit transaction and event identify one execution. The object ID
    // is resolved at the exact version recorded in that transaction's effects.
    const proof = await client
        .proof()
        .transaction(transaction)
        .object(object)
        .event(transaction, STAKING_REQUEST_EVENT_SEQUENCE)
        .build();
    const targets = proof.targets;

    assert.equal(targets.transaction, MAINNET_TRANSACTION_DIGEST);
    assert.equal(targets.objects.length, 1);
    assert.equal(targets.objects[0]?.objectId, STAKED_IOTA_OBJECT_ID);
    assert.equal(targets.events.length, 1);
    assert.equal(targets.events[0]?.transactionDigest, MAINNET_TRANSACTION_DIGEST);
    assert.equal(targets.events[0]?.eventSequence, STAKING_REQUEST_EVENT_SEQUENCE);

    console.log("Proof constructed:");
    console.log(`  checkpoint epoch: ${proof.checkpointEpoch}`);
    console.log(`  transaction targets: ${targets.transaction ? 1 : 0}`);
    console.log(`  object targets:      ${targets.objects.length}`);
    console.log(`  event targets:       ${targets.events.length}\n`);

    // Model transfer before verification. The receiver treats the entire payload
    // as untrusted input until every selected target verifies.
    const proofJSON = proof.toJSON();
    const receivedProof = Proof.fromJSON(proofJSON);
    console.log(`Serialized proof size: ${Buffer.byteLength(proofJSON)} bytes`);

    const genesis = await loadMainnetGenesis();
    const verifier = client.verifier(CommitteeResolution.fromGenesis(genesis));

    // Authenticating committee lineage through epoch 469 can take a long time.
    await verifier.verify(receivedProof);

    console.log("\nMulti-target proof verified successfully.");
    console.log("The transaction, changed object, and emitted event are authenticated by one proof.");
}
